torchani/extension/aev.cu

```diff
@@ -0,0 +1,627 @@
+
+  int maxVal = 0;
+  cudaMemcpy(&maxVal, d_out, sizeof(DataT), cudaMemcpyDefault);
+
+  CubDebugExit(g_allocator.DeviceFree(d_temp_storage));
+  //cudaFree(d_temp_storage);
+  return maxVal;
+}
+
+void
+initConsts(AEVScalarParams<float> &aev_params){  
+  int num_species = aev_params.num_species;
+  assert(num_species <= MAX_NSPECIES);
+  // precompute the aev offsets and load to constand memory
+  int *subaev_offsets = new int[num_species * num_species];
+  for(int t = 0; t < num_species; ++t){
+    int offset = 0;
+    for(int s = 0; s < num_species; s++){
+      if(t < num_species - s){
+	subaev_offsets[s * num_species + s + t] = aev_params.angular_sublength * (offset + t);
+	subaev_offsets[(s+t) * num_species + s] = aev_params.angular_sublength * (offset + t);
+      }
+      offset += num_species-s; 
+    }
+  }
+  cudaMemcpyToSymbol(csubaev_offsets, subaev_offsets,  sizeof(int) * num_species * num_species);
+  delete[] subaev_offsets;
+}
+
+
+
+//NOTE: assumes size of EtaA_t = Zeta_t = EtaR_t = 1  
+template<typename ScalarRealT = float>
+void
+cuComputeAEV(torch::Tensor coordinates_t,
+	     torch::Tensor species_t,
+	     ScalarRealT Rcr,
+	     ScalarRealT Rca,	     
+	     torch::Tensor EtaR_t,
+	     torch::Tensor ShfR_t,
+	     torch::Tensor EtaA_t,
+	     torch::Tensor Zeta_t,
+	     torch::Tensor ShfA_t,
+	     torch::Tensor ShfZ_t,
+	     torch::Tensor aev_t,
+	     int num_species){
+  const int n_molecules = species_t.size(0);
+  const int max_natoms_per_mol = species_t.size(1);
+  //std::cout << "Running cuComputeAEV with " << n_molecules << "x" << max_natoms_per_mol << "\n";
+
+  AEVScalarParams<float> aev_params;
+  aev_params.Rca = Rca;
+  aev_params.Rcr = Rcr;
+  aev_params.num_species = num_species;
+
+  aev_params.radial_sublength = EtaR_t.size(0) * ShfR_t.size(0); 
+  aev_params.radial_length = aev_params.radial_sublength * num_species;
+
+  aev_params.angular_sublength = EtaA_t.size(0) * Zeta_t.size(0) * ShfA_t.size(0) * ShfZ_t.size(0); 
+  aev_params.angular_length = aev_params.angular_sublength * (num_species * (num_species + 1)/2);
+
+  if(EtaR_t.size(0) != 1 || EtaA_t.size(0) != 1 || Zeta_t.size(0) != 1){
+    std::cerr << "cuda extension is currently not supported for the specified configuration\n";
+    exit(1);
+  }
+  
+  // precompute the aev offsets and load to constand memory
+  initConsts(aev_params);
+
+  // buffer to store all the pairwise distance (Rij)
+  PairDist<float> *d_Rij = 0;
+  auto total_natom_pairs = n_molecules * max_natoms_per_mol * max_natoms_per_mol;
+  CubDebugExit(g_allocator.DeviceAllocate((void**)&d_Rij, sizeof(PairDist<float>) * total_natom_pairs));
+  //cudaMalloc((void **)&d_Rij, sizeof(PairDist<float>) * total_natom_pairs);
+
+  // init all Rij to inf
+  PairDist<float> init;
+  init.Rij = std::numeric_limits<float>::infinity();
+  thrust::fill(thrust::device, d_Rij, d_Rij + total_natom_pairs, init);

```

Comment:
I think we should run this on stream `c10::cuda::getCurrentCUDAStream()`, instead of the default stream.

Reply:
Working at https://github.com/akkamesh/torchani/pull/1